pageserver/src/tenant/storage_layer.rs

```diff
@@ -336,6 +311,952 @@ impl LayerAccessStats {
+            // this is sadly needed because of task_mgr::shutdown_tasks, otherwise we cannot
+            // block tenant::mgr::remove_tenant_from_memory.
+            let this = self.clone();
+            crate::task_mgr::spawn(
+                &tokio::runtime::Handle::current(),
+                TaskKind::RemoteDownloadTask,
+                Some(self.desc.tenant_id),
+                Some(self.desc.timeline_id),
+                &task_name,
+                false,
+                async move {
+                    let client = timeline
+                        .remote_client
+                        .as_ref()
+                        .expect("checked above with can_ever_evict");
+                    let result = client
+                        .download_layer_file(
+                            &this.desc.filename(),
+                            &crate::tenant::remote_timeline_client::index::LayerFileMetadata::new(
+                                this.desc.file_size,
+                            ),
+                        )
+                        .await;
+
+                    match result {
+                        Ok(size) => {
+                            timeline.metrics.resident_physical_size_gauge.add(size);
+                            let _ = tx.send(());
+                        }
+                        Err(e) => {
+                            // TODO: the temp file might still be around, metrics might be off
+                            tracing::error!("layer file download failed: {e:?}",);
+                        }
+                    }
+
+                    Ok(())
+                }
+                .in_current_span(),
+            );
+            if rx.await.is_err() {
+                return Err(anyhow::anyhow!("downloading failed, possibly for shutdown"));
+            }
+            // FIXME: we need backoff here so never spiral to download loop
+            anyhow::ensure!(
+                self.needs_download()
+                    .await
+                    .context("test if downloading is still needed")?
+                    .is_none(),
+                "post-condition for downloading: no longer needs downloading"
+            );
+        } else {
+            // the file is present locally and we could even be running without remote
+            // storage
+        }
+
+        // the assumption is that we own the layer residentness, no operator should go in
+        // and delete random files. this would be evident when trying to access the file
+        // Nth time (N>1) while having the VirtualFile evicted in between.
+        //
+        // we could support this by looping on NotFound from the layer access methods, but
+        // it's difficult to implement this so that the operator does not delete
+        // not-yet-uploaded files.
+
+        let res = Arc::new(DownloadedLayer {
+            owner: Arc::downgrade(self),
+            kind: tokio::sync::OnceCell::default(),
+        });
```

Comment:
Yeah I was thinking of making this so that you could always go and delete files, but then realized the problem of "making sure operator does not delete not-yet-uploaded files". I guess we *could* sidestep that with keeping an fd open. Or hope eviction works well enough and not support this.

This is related to the state tracking.

The proposed comment is already on the type of DownloadedLayer.